Enable app class independent usage

bebuch opened the pull request:
This is a pure extension and fully backwards compatible. (Unless I made a mistake. 😉)

I have two requirements for the library that are not currently possible:

1. I want to use it as a pre-compiled library while still being able to choose my `QXxxApplication` class at build time.
2. I need to be able to decide at runtime whether to use the single instance stuff (without starting a server and so on).

This PR allows both by making the `SingleApplication` class an object that is not derived from `QXxxApplication`. On the user side, you need to instantiate a separate `SingleApplication` object after your `QXxxApplication` object. I've added a new example `separate_object` which mirrors `sending_arguments` but uses `SingleApplication` as a separate object.

Of course, the better approach would be to put the implementation into a separate class with an appropriate name and then use that to implement `SingleApplication` as it currently is. I would be delighted if this were implemented for 4.x. Then you could also use the functionality with installed CMake Config and not just as a Git submodule.

If you are willing to merge this, I would also create a PR with CMake install for CMake Config files in “FreeStanding” mode.

If you don't want to merge this, but are willing to merge the “better approach” mentioned, then I can also implement a corresponding PR for 3.x.

@itay-grudev What do you think?

itay-grudev commented:
I get it. It looks OK as a concept. Isn't there some extra work required for example calls like this one won't work when it doesn't inherit from `QCoreApplication`:

https://github.com/itay-grudev/SingleApplication/blob/494772e98cef0aa88124f154feb575cc60b08b38/singleapplication.cpp#L171

itay-grudev commented:
If you it is entirely backwards compatible I don't mind merging it as a `v3.6` as I don't currently have time time to finish the `v4` implementation that doesn't use `QSharedMemory`.

bebuch commented:
Thanks for the fast review, sounds good!

> I get it. It looks OK as a concept. Isn't there some extra work required for example calls like this one won't work when it doesn't inherit from `QCoreApplication`:
> 
> https://github.com/itay-grudev/SingleApplication/blob/494772e98cef0aa88124f154feb575cc60b08b38/singleapplication.cpp#L171

If I understand this correctly you call the C standard library funktion [`::exit`](https://en.cppreference.com/w/c/program/exit) here. According to [Stackoverflow](https://stackoverflow.com/questions/38724788/what-is-the-difference-between-exit-and-stdexit-in-c), this is equivalent to the C++ [`std::exit`](https://en.cppreference.com/w/cpp/utility/program/exit) pendant.

To my knowledge, it makes no difference whether I call this from a `QCoreApplication` class or from somewhere else. Am I wrong there? 🤔

bebuch commented:
I've fixed the bug with the preprocessor string compartion. Looks like this is not possible. Instead I've added a new file `FreeStandingSingleApplication` that works as the fake `QXxxApplication` base class.

In the new `separate_object` example I've renamed the executable to `separate_object`.

I've successfully tested all 5 examples locally on Windows with clang. I'm going to test with MSVC and on Linux with GCC as well.

Please run the CI pipelines to validate that my changes didn't break anything.

I'm going to add a new CI pipeline step for the new example and add documentation on how to use the free standing mode.

bebuch commented:
I added CI steps for the new `separate_object` example with QMake and CMake.

I added a CMake install in free standing mode that generates CMake config files and removes the `QAPPLICATION_CLASS` macro from the installed header file. I added user doc to readme and notes to changelog.

A notable change is, that I added the library version (3.6.0!) to the `CMakeLists.txt` statement `project`! Thats required to generate Config files that can test version compatability.

From my side this is ready for merge now. Let me know I you want changes or more explenation 😉

bebuch commented:
```
ld: warning: ignoring file 'QtCore.framework/Versions/5/QtCore': found architecture 'x86_64', required architecture 'arm64'
ld: warning: ignoring file 'QtNetwork.framework/Versions/5/QtNetwork': found architecture 'x86_64', required architecture 'arm64'
```

This looks like a broken build environment to me. 🤔

bebuch commented:
Okay, from what I've found MacOS runners are now arm64 only. X86_64 is no longer supported with macos-latest. According to an fmt issue macos-11 is also no longer supported.

https://github.com/jurplel/install-qt-action

If I understand the action file right, than this is the repo responsible for installing Qt. In this case it's probably a bug on their side with the Qt5 install on current (arm64) MacOS.

Maybe it's the best to temporarily disable MacOS in the CI. What do you think?

bebuch commented:
I've made a new PR #204 to fix the MacOS CI build. If that succeeds and got merged, I will rebase this PR and we can continue 😅

bebuch commented:
I rebased this on my CI-fix PR #204 and the CI succeeds now.

https://github.com/bebuch/SingleApplication/actions/runs/12872062233/

This is ready for merge from my side. (If you accept #204 as well.)

itay-grudev commented on .github/workflows/main.yml:
```diff
@@ -80,31 +80,48 @@ jobs:
           cmake --build .
 
       - name: Build basic example with QMake
+        if: ${{ !contains(matrix.platform, 'macos') }}
```
Why are these ignored in MacOS?

itay-grudev replied:
I imagine only the Windows example should be ignored.
